File: solutions/java/rest-api/1/src/main/java/RestApi.java
import java.util.HashMap;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONObject;

class RestApi {
    private final Map<String, Map<String, Double>> netBorrow = new HashMap<>();

    RestApi(User... users) {
        for (var user: users) {
            netBorrow.putIfAbsent(user.name(), new HashMap<>());
            var userEntry = netBorrow.get(user.name());
            for (var iou: user.owes()) {
                userEntry.put(iou.name, userEntry.getOrDefault(iou.name, 0.0) + iou.amount);
            }
            for (var iou: user.owedBy()) {
                userEntry.put(iou.name, userEntry.getOrDefault(iou.name, 0.0) - iou.amount);
            }
        }
        setOff();
    }

    private void setOff() {
        netBorrow.entrySet().forEach(netBorrowEntry -> {
            netBorrowEntry.getValue().entrySet().removeIf(e -> e.getValue() == 0.0);
        });
    }

    String get(String url) {
        return switch (url) {
            case "/users" -> getAllUsers().toString();
            case "/add" -> throw new UnsupportedOperationException("/add is not a GET endpoint");
            case "/iou" -> throw new UnsupportedOperationException("/iou is not a GET endpoint");
            default -> throw new IllegalArgumentException("Bad URL: " + url);
        };
    }

    String get(String url, JSONObject payload) {
        return switch (url) {
            case "/users" -> getUsers(payload.optJSONArray("users")).toString();
            case "/add" -> throw new UnsupportedOperationException("/add is not a GET endpoint");
            case "/iou" -> throw new UnsupportedOperationException("/iou is not a GET endpoint");
            default -> throw new IllegalArgumentException("Bad URL: " + url);
        };
    }

    String post(String url, JSONObject payload) {
        return switch (url) {
            case "/users" -> throw new UnsupportedOperationException("/users is not a POST endpoint");
            case "/add" -> addUser(payload.optString("user")).toString();
            case "/iou" -> addBorrow(
                payload.optString("borrower"), payload.optString("lender"), payload.optDouble("amount")
            ).toString();
            default -> throw new IllegalArgumentException("Bad URL: " + url);
        };
    }

    private JSONObject singleUser(String borrower, Map<String, Double> lenders) {
        var user = new JSONObject().put("name", borrower);
        var owes = new JSONObject();
        var owedBy = new JSONObject();
        double[] balance = {0.0};
        lenders.forEach((lender, amount) -> {
            balance[0] -= amount;
            if (amount < 0.0) {
                owedBy.put(lender, -amount);
            }
            if (amount > 0.0) {
                owes.put(lender, amount);
            }
        });
        return user.put("owes", owes).put("owedBy", owedBy).put("balance", balance[0]);
    }

    JSONObject getAllUsers() {
        var users = new JSONArray();
        netBorrow.forEach((borrower, lenders) -> {
            users.put(singleUser(borrower, lenders));
        });
        return new JSONObject().put("users", users);
    }

    JSONObject getUsers(JSONArray usernames) {
        var users = new JSONArray();
        usernames.forEach(borrowerObj -> {
            String borrower = (String) borrowerObj;
            var lenders = netBorrow.get(borrower);
            if (lenders != null) {
                users.put(singleUser(borrower, lenders));
            }
        });
        return new JSONObject().put("users", users);
    }

    JSONObject addUser(String username) {
        netBorrow.putIfAbsent(username, new HashMap<>());
        return singleUser(username, netBorrow.get(username));
    }

    JSONObject addBorrow(String borrower, String lender, double amount) {
        netBorrow.putIfAbsent(borrower, new HashMap<>());
        netBorrow.putIfAbsent(lender, new HashMap<>());
        var borrowerData = netBorrow.get(borrower);
        var lenderData = netBorrow.get(lender);
        borrowerData.put(lender, borrowerData.getOrDefault(lender, 0.0) + amount);
        lenderData.put(borrower, lenderData.getOrDefault(borrower, 0.0) - amount);
        setOff();
        if (borrower.compareTo(lender) < 0) {
            return getUsers(new JSONArray().put(borrower).put(lender));
        } else {
            return getUsers(new JSONArray().put(lender).put(borrower));
        }
    }
}

File: solutions/python/rest-api/1/rest_api.py
from json import dumps, loads
from typing import Literal

GET_URL_t = Literal["/users"]
POST_URL_t = Literal["/add", "/iou"]

class RestAPI:
    def __init__(self, database: dict = {}) -> None:
        self._db = {user.get('name'): user \
                    for user in database.get('users', [])}

    def get(self, url: GET_URL_t, payload: str | None = None) -> str:
        match url:
            case "/users":
                return self._get_users(payload) \
                    if payload else self._get_all_users()
            case _:
                raise ValueError(f'Invalid GET URL: {url}')

    def post(self, url: POST_URL_t, payload: str = "{}") -> str:
        match url:
            case "/add":
                return self._add_user(payload)
            case "/iou":
                return self._update_iou(payload)
            case _:
                raise ValueError(f'Invalid POST URL: {url}')

    def _get_all_users(self) -> str:
        return dumps({'users': [*self._db.values()]})

    def _get_users_from_list(self, names: list) -> str:
        return dumps({'users': [self._db.get(name) for name in names if name in self._db]})
    
    def _get_users(self, payload: str) -> str:
        return self._get_users_from_list(loads(payload).get('users', []))

    def _add_user(self, payload: str) -> str:
        name: str = loads(payload).get('user', '')
        user: dict = RestAPI._default_user(name)
        return dumps(self._db.setdefault(name, user))

    def _update_iou(self, payload: str) -> str:
        pl: dict = loads(payload)
        lender_name: str = pl.get('lender', '')
        borrower_name: str = pl.get('borrower', '')
        amount: float = pl.get('amount', 0.0)
        lender: dict = self._db.get(lender_name, RestAPI._default_user(lender_name))
        borrower: dict = self._db.get(borrower_name, RestAPI._default_user(borrower_name))
        lender['balance'] += amount
        borrower['balance'] -= amount
        lend: float = lender['owed_by'].pop(borrower_name, 0.0) - lender['owes'].pop(borrower_name, 0.0)
        borrower['owed_by'].pop(lender_name, 0.0)
        borrower['owes'].pop(lender_name, 0.0)
        lend += amount
        if lend > 0:
            lender['owed_by'][borrower_name] = lend
            borrower['owes'][lender_name] = lend
        if lend < 0:
            borrower['owed_by'][lender_name] = -lend
            lender['owes'][borrower_name] = -lend
        self._db |= {lender_name: lender, borrower_name: borrower}
        return self._get_users_from_list(sorted([lender_name, borrower_name]))
    
    @staticmethod
    def _default_user(name: str) -> dict:
        return {'name': name, 'owes': {}, 'owed_by': {}, 'balance': 0.0}